pkg/signature/algorithm_registry.go

```diff
@@ -0,0 +1,239 @@
+		if detail.knownAlgorithm == knownSignatureAlgorithm {
+			return &detail, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find algorithm details for known signature algorithm: %s", knownSignatureAlgorithm)
+}
+
+// NewAlgorithmRegistryConfig creates a new AlgorithmRegistryConfig for a set of permitted signature algorithms.
+func NewAlgorithmRegistryConfig(algorithmConfig []v1.KnownSignatureAlgorithm) (*AlgorithmRegistryConfig, error) {
+	permittedAlgorithms := make([]AlgorithmDetails, 0, len(algorithmDetails))
+	for _, algorithm := range algorithmConfig {
+		a, err := GetAlgorithmDetails(algorithm)
+		if err != nil {
+			return nil, err
+		}
+		permittedAlgorithms = append(permittedAlgorithms, a)
+	}
+	return &AlgorithmRegistryConfig{permittedAlgorithms: permittedAlgorithms}, nil
+}
+
+// IsAlgorithmPermitted checks whether a given public key/hash algorithm combination is permitted by a registry config.
+func (registryConfig AlgorithmRegistryConfig) IsAlgorithmPermitted(key crypto.PublicKey, hash crypto.Hash) bool {
+	for _, algorithm := range registryConfig.permittedAlgorithms {
+		if algorithm.checkKey(key) && algorithm.checkHash(hash) {
+			return true
+		}
+	}
+	return false
+}
+
+// GetAlgorithmDetails retrieves a set of details for a given v1.KnownSignatureAlgorithm flag that allows users to
+// introspect the public key algorithm, hash algorithm and more. Only the algorithms that are permitted by the registry
+// are returned.
+func (registryConfig AlgorithmRegistryConfig) GetAlgorithmDetails(signatureAlgorithm v1.KnownSignatureAlgorithm) (AlgorithmDetails, error) {
+	for _, detail := range registryConfig.permittedAlgorithms {
+		if detail.GetSignatureAlgorithm() == signatureAlgorithm {
+			return detail, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find permitted algorithm details for known signature algorithm: %s", signatureAlgorithm)
+}
+
+// ListPermittedAlgorithms returns a list of permitted algorithms for a given registry config.
+func (registryConfig AlgorithmRegistryConfig) ListPermittedAlgorithms() []v1.KnownSignatureAlgorithm {
+	permittedAlgorithms := make([]v1.KnownSignatureAlgorithm, 0, len(registryConfig.permittedAlgorithms))
+	for _, algorithm := range registryConfig.permittedAlgorithms {
+		permittedAlgorithms = append(permittedAlgorithms, algorithm.GetSignatureAlgorithm())
+	}
+	return permittedAlgorithms
+}
+
+// HasAlgorithmDetails checks whether a given v1.KnownSignatureAlgorithm flag is permitted by the registry.
+func (registryConfig AlgorithmRegistryConfig) HasAlgorithmDetails(signatureAlgorithm v1.KnownSignatureAlgorithm) bool {
+	_, err := registryConfig.GetAlgorithmDetails(signatureAlgorithm)
+	return err == nil
+}
+
+// FormatSignatureAlgorithmFlag formats a v1.KnownSignatureAlgorithm to a string that conforms to the conventions of CLI
+// arguments that are used for Sigstore services.
+func FormatSignatureAlgorithmFlag(algorithm v1.KnownSignatureAlgorithm) (*string, error) {
```

Comment:
Any reason for a string pointer over a string?

Reply:
Just so I don't have to construct a string in the error case. I can just create an empty one though if that's preferable.

Reply:
[1d74008](https://github.com/sigstore/sigstore/pull/1601/commits/1d74008e9c24c141132f550f38e158455e46c259)